### PR TITLE
Enforce minimum service factor limits (take 2)

### LINKED ISSUE #320: Invalid values for minimum service factor in `test_minimum_service_factor`
As established in #304, `MinimumServiceFactor` should be between 0 and 1. However, the `test_minimum_service_factor` test uses invalid values >1, and still passes

Ideally we shouldn't be in a position where the model can be run with invalid values, however at the very least this test should be modified to use valid values.

It's also worth making sure this test is checking the correct thing. It seems to be comparing the absolute value of supply against the `MinimumServiceFactor`, however my understanding is that `MinimumServiceFactor` relates to relative supply compared to the theoretical maximum (hence why it needs to be between 0 and 1). Or maybe I'm misunderstanding something...

### PR DESCRIPTION
**Pull request by alexdewar**
Let's try this again. It turns out the fix for my previous mistake was really trivial (5fb3ff5824ae034bd11955006001a5dde321d897). Sorry about all the noise!

Fixes #320.

### CODE REVIEW

File: src/muse/readers/csv.py [hunk 1]
```diff
@@ -129,7 +133,7 @@ def read_technodata_timeslices(filename: Union[Text, Path]) -> xr.Dataset:
     data = csv[csv.technology != "Unit"]
 
     data = data.apply(lambda x: pd.to_numeric(x, errors="ignore"))
-    data = check_utilization_not_all_zero(data, filename)
+    check_utilization_not_all_zero(data, filename)
```
Comment (dalonsoa) [inline src/muse/readers/csv.py:136]: in principle, the `technodata_timeslices` file also has an `utilization_factor` column that will need validating. See https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/enforce_minimum_service_factor_limits/src/muse/data/example/default_timeslice/technodata/power/TechnodataTimeslices.csv

File: tests/test_minimum_service.py [hunk 2]
```diff
@@ -1,36 +1,37 @@
...
+    permutations((np.linspace(0, 1, 6), [0] * 6)),
 )
-def test_minimum_service_factor(tmpdir, minimum_service_factor, process_name):
+def test_minimum_service_factor(tmpdir, minimum_service_factors):
     import pandas as pd
     from muse import examples
     from muse.mca import MCA
 
     sector = "power"
+    processes = ("gasCCGT", "windturbine")
```
Comment (dalonsoa) [inline tests/test_minimum_service.py:34]: Concrete values of the service factor aside, this does not check the same scenarios as before. Before there were 4 tests being launched, and now they are only 2. Not that the previous 4 cases were necessary, but just flagging this.

**Review by dalonsoa — CHANGES_REQUESTED** [review]
The changes look ok, but I've a couple of comments/concerns, and a request.

**Review by tsmbland — CHANGES_REQUESTED** [review]
Looks good, although I think we also need to have a similar check for `MinimumServiceFactor` in the `read_technodata_timeslices` function. While we're at it, it would also be good to have a similar check for `UtilizationFactor` (i.e. it needs to be between 0 and 1). And we could also check that `UtilizationFactor` is always >= `MinimumServiceFactor`.

Also, it might be nice to have a some tests with invalid inputs to make sure these checks fail when they should.

**Comment by alexdewar** [comment]
All good points @tsmbland!

File: tests/test_minimum_service.py (hunk 2 above)
Comment (alexdewar) [reply tests/test_minimum_service.py:34]: There should still be 4... `permutations()` returns the inputs in 2 different orders

**Comment by alexdewar** [comment]
Ok, I think I've made all the changes that were suggested and I've added a bunch more tests. I'm glad I had a more thorough look at this because it turns out my original fix didn't actually work :worried:

**Review by dalonsoa — APPROVED** [review]
Looks good! Way more comprehensive validation than before.

**Comment by dalonsoa** [comment]
But tests are failing...

**Comment by alexdewar** [comment]
Sorry :disappointed:. I pushed this at the end of the day and figured it was all fine (the tests I was working on were passing!). I'll take another look. Hopefully it's just something trivial!

**Comment by alexdewar** [comment]
Converting to draft until I fix the test failures (see also #362)

File: tests/test_readers.py [hunk 3]
```diff
@@ -410,3 +412,150 @@ def test_read_trade_technodata(tmp_path):
...
+        mock.assert_called_once_with(df, "file.csv")
+    min_service_factor_mock.assert_not_called()
+    utilization_below_min_mock.assert_not_called()
+
+
+@patch("muse.readers.csv._check_utilization_in_range")
+@patch("muse.readers.csv._check_utilization_not_all_zero")
+@patch("muse.readers.csv._check_utilization_not_below_minimum")
+@patch("muse.readers.csv._check_minimum_service_factors_in_range")
+def test_check_utilization_and_minimum_service_factors_fail_missing_utilization(*mocks):
```
Comment (tsmbland) [inline tests/test_readers.py:547]: I don't think you need the mocks here. Otherwise, everything looks good!

**Review by tsmbland — APPROVED** [review]

**Pull request merged**